### PR TITLE
Partial DIP66, interoperation alias this and inheritance.

### PR DESCRIPTION
**Pull request by IgorStepanov**
This PR partially revives #3998, which implements [DIP66](https://wiki.dlang.org/DIP66) (Three years have passed :))
It doesn't introduce multiple alias this (for the same class or struct)  yet, however it implements interoperability with inheritance. Now we may implement multiple alias this through multiple inheritance of interfaces and properties.
I deliberately moved multiple alias this out of bounds of this PR, because I want to reduce atomic changes size.
Also this PR doesn't touch interoperation alias this and tuples (It is goal of the next PR).
If (When) this PR will be approved, I easily introduce multiple alias this without complex changes.

### CODE REVIEW

**Comment by JinShil** [comment]
Hey! Welcome back, @IgorStepanov

**Comment by IgorStepanov** [comment]
I need help. What happened with CircleCI test?
```
/home/circleci/dmd/generated/linux/release/64/backend.a(cgcod.o):(.data.rel+0x0): multiple definition of `cdxxx'
/home/circleci/dmd/generated/linux/release/64/backend.a(optabgen.o):(.bss+0x800): first defined here
``` 
I haven't touch backend in this PR.

**Comment by ibuclaw** [comment]
There's still a lot here so will take time to sift through. Commit message itself could be improved also. If there's anything change that can be applied independently, splitting it off into a separate commit also helps.

**Comment by JinShil** [comment]
> What happened with CircleCI test?

I'm seeing the same error here:  https://github.com/dlang/dmd/pull/8228  Looks like there's something unrelated at play.

**Comment by IgorStepanov** [comment]
>Commit message itself could be improved also.

Done

>If there's anything change that can be applied independently, splitting it off into a separate commit also helps.

Unfortunately I can't find a part that can be commited separately. All changes looks dependent on each other. :(

**Comment by IgorStepanov** [comment]
@ibuclaw ping.

**Comment by JinShil** [comment]
Can you add some more tests to get coverage to 100%?

**Comment by wilzbach** [comment]
@IgorStepanov this needs to pass on the Project Tester (D wants to be a stable language and we don't want silent regressions).
I see that a few projects don't build anymore, e.g. with "unable to unambiguously resolve".
In those case we should fallback to the previous behaviour and only issue a deprecation message, s.t. projects have enough time to upgrade their code and it's not red overnight.

**Comment by IgorStepanov** [comment]
@wilzbach May I force changes to failed projects instead?
I can't simply fallback to the previous behaviour in some cases. This PR completely changes the approach to the alias this resolution. 
I can convert error "unable to unambiguously resolve" to depreaction messages and get the first choice.
BTW: in `ae` test fails on a hack, which needs to overcome [issue5973](https://issues.dlang.org/show_bug.cgi?id=5973), which fixes by this PR.

However there are two tests (`sociomantic-tsunami/ocean` and `kaleidicassociates/excel-d`) which may represent show possible bugs of this PR. I'll see them tomorrow.

**Comment by wilzbach** [comment]
> May I force changes to failed projects instead?

Kinda. First of, in the end the project tester needs to be green. That being said, it's ok to upgrade the projects (they aren't set in stone, but their latest git tag is fetched) and we don't have a general policy of when it's ok/not ok to upgrade the projects, but IIRC Walter and Razvan ran into a similar issues a few months ago and it was decided that's ok to modify the project directly.
Anyhow, bear in mind that people really hate breaking changes nowadays.

> I can't simply fallback to the previous behaviour in some cases. This PR completely changes the approach to the alias this resolution.
> I can convert error "unable to unambiguously resolve" to depreaction messages and get the first choice.

Yes that's what I was thinking about too. I just hope that's equivalent to what DMD is doing at the moment.

> BTW: in ae test fails on a hack, which needs to overcome issue5973, which fixes by this PR.

This does sound like a good example where it's okay to modify the downstream projects.
However, I'm pretty sure that @CyberShadow would only accept a PR if it doesn't break his project with the currently released master (you might get around this with a `static if` tough).

> However there are two tests (sociomantic-tsunami/ocean and kaleidicassociates/excel-d) which may represent show possible bugs of this PR.

That's why the project tester was created ;-)

**Comment by jacob-carlborg** [comment]
It’s perfectly fine to fix downstream projects, but it’s very likely that other projects not part of our test suite is affected as well.

**Comment by IgorStepanov** [comment]
@jacob-carlborg Is it correct to generate warning in conflict case? 
I generate warning instead of error for `s-ludwig/taggedalgebraic` case.
However test is failed.

**Comment by Geod24** [comment]
@IgorStepanov : Deprecations should be generated, not warnings, as warnings are in many projects treated as errors.

**Comment by WalterBright** [comment]
I'm having a hard time with this. It's 3000 lines of changes, most of which are substantive. There's no guide to it, no description of what the implementation idea is. One is left with trying to piece something together from the diffs spread out over 20 files.

**Comment by IgorStepanov** [comment]
@WalterBright Unfortunately, I have no idea, how to split this PR to small parts. I think, if we want to introduce multiple alias this, we must go this way.
If you have any idea, how can I help to review this, please say.
Now, I removed all places (which I found), where this PR breaks existing code.
Looks like it passes all jenkins tests, except 7 vibe.d based projects. However, this failure is irrelevant with my PR. ([See](https://ci.dlang.io/blue/organizations/jenkins/dlang-org%2Fdmd/detail/PR-7528/4/pipeline/259) another projects.)

**Comment by JinShil** [comment]
@IgorStepanov,  @WalterBright would have to speak for himself what would make this reviewable for him, but what would help me is to have some documentation accompanying this that describes the implementation -- as @WalterBright said, "a guide to it".  Detailed code comments would help a lot.  A detailed overview as the first comment of this PR would also be quite welcome.  Basically, just help others see the mental model that you have.  From there, questions would probably ensue, and as long as you are attentive to those questions, it probably wouldn't be long until this were merged.

And, we would really like 100% test coverage.

File: src/dmd/dcast.d
Comment (JinShil) [inline src/dmd/dcast.d]: I see the above block of code repeated 3 times in this file.  Perhaps you could isolate this in nested function and then add a documentation comment to describe it.  I think that would help the review.

File: src/dmd/aliasthis.d
Comment (JinShil) [inline src/dmd/aliasthis.d]: I really don't like the `function2` naming convention used in DMD.  I know such names are used elsewhere in DMD.  Perhaps add a `bool` parameter to `resolveAliasThis`.  It's just a suggestion, though; I'd still merge this accept this even if no changes were made.

File: src/dmd/expressionsem.d
Comment (JinShil) [inline src/dmd/expressionsem.d]: "On recursion done in iterateAliasThis" -- I can't understand this sentence. Is there a typo somwhere?

File: src/dmd/expressionsem.d
Comment (JinShil) [inline src/dmd/expressionsem.d]: I see this error message displayed multiple times in this PR, with accompanying `iterateAliasThis`.  Is it possible to reduce the code duplication in a nested function or something similar?  With it located in one logical block, it would also open the opportunity to document it.  It may not be worth it though; it's just a suggestion.

File: src/dmd/expressionsem.d
Comment (JinShil) [inline src/dmd/expressionsem.d]: Again, I'm seeing code duplication here.  Is there a way to reduce that and supply some accompanying documentation for it?

**Comment by JinShil** [comment]
Overall, IMO, this is looking really good.  It's just a difficult PR to review.

File: src/dmd/opover.d
Comment (IgorStepanov) [reply src/dmd/opover.d]: I may create a function with ton of arguments which combine `iterateAliasThis` and `enforceOneResult`, but I don't want to do that, because `iterateAliasThis` has a default values for the last arguments and `enforceOneResult` has a variadic arguments.
And they do independent things:
`iterateAliasThis` collects candidates to resolve,
`enforceOneResult` checks that there is only one candidate and raises an error otherwise.
And `enforceOneResult` and take a different error messages. They similar but not the same.
What do you think?

File: src/dmd/aliasthis.d
Comment (IgorStepanov) [reply src/dmd/aliasthis.d]: fixed

File: src/dmd/expressionsem.d
Comment (IgorStepanov) [reply src/dmd/expressionsem.d]: fixed

File: src/dmd/mtype.d
Comment (IgorStepanov) [reply src/dmd/mtype.d]: fixed

File: src/dmd/dcast.d
Comment (IgorStepanov) [reply src/dmd/dcast.d]: fixed

**Comment by IgorStepanov** [comment]
@JinShil I understood the request for additional explanations.
I'll created it at the following week.

**Comment by IgorStepanov** [comment]
How can I see details of the failed test?

File: src/dmd/mtype.d
Comment (JinShil) [inline src/dmd/mtype.d]: You may want to explicitly return `MOD`.  See also https://github.com/dlang/dmd/pull/8495

File: src/dmd/dcast.d
Comment (JinShil) [inline src/dmd/dcast.d]: I think this function should be `static` also.

File: src/dmd/opover.d
Comment (JinShil) [reply src/dmd/opover.d]: If you can't find a way to consolidate the context so you don't need to create a function with too many arguments, I'd say it's not worth it, but I hope there's a way.

File: src/dmd/dcast.d
Comment (IgorStepanov) [reply src/dmd/dcast.d]: Nope, because it modifies local variables of outer function.

**Comment by JinShil** [comment]
> How can I see details of the failed test?

The BuildKite logs aren't public yet.  Here's what I see.  It doesn't look related to this PR.

```
+ cd ggplotd
+ case "$REPO_FULL_NAME" in
+ sed -i 's|auto seed = unpredictableSeed|auto seed = 54321|' source/ggplotd/example.d
+ use_travis_test_script
+ bash
+ /var/lib/buildkite-agent/builds/dime-2/dlang/dmd/distribution/buildkite/travis_get_script
Fetching cairod 0.0.1-alpha.3+1.10.2 (getting selected version)...
Fetching windows-headers 1.0.1 (getting selected version)...
Fetching derelict-util 2.1.0 (getting selected version)...
Fetching painlesstraits 0.2.0 (getting selected version)...
Fetching x11 1.0.20 (getting selected version)...
Fetching derelict-ft 1.0.2 (getting selected version)...
Fetching dstats 1.0.5 (getting selected version)...
Fetching color 0.0.7 (getting selected version)...
Fetching gtk-d 3.3.1 (getting selected version)...
Fetching dunit 1.0.14 (getting selected version)...
Generating test runner configuration 'ggplotd-test-default' for 'default' (library).
Performing "unittest" build using dmd for x86_64.
derelict-util 2.1.0: building configuration "library"...
derelict-ft 1.0.2: building configuration "library"...
x11 1.0.20: building configuration "library"...
../.dub/packages/x11-1.0.20/x11/source/x11/extensions/XI.d(3,8): Deprecation: Symbol `object.string` is not visible from module `XI` because it is privately imported in module `string`
cairod 0.0.1-alpha.3+1.10.2: building configuration "static-custom-posix"...
dstats 1.0.5: building configuration "library"...
dunit 1.0.14: building configuration "library"...
painlesstraits 0.2.0: building configuration "library"...
ggplotd 1.1.6: building configuration "ggplotd-test-default"...
Linking...
Running ./ggplotd-test-default 
Random seed MCMC: 54321
Random seed MCMC: 54321
All unit tests have been run successfully.
Generating test runner configuration 'ggplotd-test-ggplotd-gtk' for 'ggplotd-gtk' (library).
Performing "unittest" build using dmd for x86_64.
derelict-util 2.1.0: building configuration "library"...
derelict-ft 1.0.2: building configuration "library"...
x11 1.0.20: building configuration "library"...
../.dub/packages/x11-1.0.20/x11/source/x11/extensions/XI.d(3,8): Deprecation: Symbol `object.string` is not visible from module `XI` because it is privately imported in module `string`
cairod 0.0.1-alpha.3+1.10.2: building configuration "static-custom-posix"...
dstats 1.0.5: building configuration "library"...
dunit 1.0.14: building configuration "library"...
gtk-d:gtkd 3.3.1: building configuration "library"...
painlesstraits 0.2.0: building configuration "library"...
ggplotd 1.1.6: building configuration "ggplotd-test-ggplotd-gtk"...
Killed
dmd failed with exit code 137.
Error: The command exited with status 2
```

File: src/dmd/mtype.d
Comment (IgorStepanov) [reply src/dmd/mtype.d]: Done

**Comment by IgorStepanov** [comment]
Remaining 2 fails is not relevant with my PR, I hope.
```
/usr/bin/ld: error in /tmp/.dub/build/tests-~master/application-debug-linux.posix-x86_64-dmd_2081-1E2A9CC457960246BDE6C044D4450043/tests.o(.eh_frame); no .eh_frame_hdr table will be created.
/usr/bin/ld: /tmp/.dub/build/tests-~master/application-debug-linux.posix-x86_64-dmd_2081-1E2A9CC457960246BDE6C044D4450043/tests.o: invalid string offset 1146189899 >= 96137 for section `.strtab'
/usr/bin/ld: final link failed: Nonrepresentable section on output
```
Why does this happen?

**Comment by wilzbach** [comment]
> Remaining 2 fails is not relevant with my PR, I hope.

Yeah that's unrelated as well (it happened here too: https://ci.dlang.io/blue/organizations/jenkins/dlang-org%2Fdmd/detail/PR-8499/6/pipeline). Not sure why it happens though.

**Comment by wilzbach** [comment]
> Remaining 2 fails is not relevant with my PR, I hope.

Yeah that's unrelated as well (it happened here too: https://ci.dlang.io/blue/organizations/jenkins/dlang-org%2Fdmd/detail/PR-8499/6/pipeline). Not sure why it happens though.

**Comment by marler8997** [comment]
Very much looking forward to this feature.  Thanks for reviving this @IgorStepanov !

**Comment by ibuclaw** [comment]
@IgorStepanov - just thinking out loud, is it not possible to better encapsulate alias this implementation so we're not making changes all over the place?

I really do feel like we should pull up the handbrake here and acknowledge that alias this is support should not be made more complex than what it already is. I think first steps would be to move as much - if not all - logic to aliasthis.d, then make the necessary improvements this pr introduces.

I'll highlight some parts which show examples of what I mean about the complexity.

File: src/dmd/dcast.d
Comment (ibuclaw) [inline src/dmd/dcast.d]: Here, introduction of a complex block to handle alias this resolving where it was once only a call to `resolveAliasThis()`.

File: src/dmd/dcast.d
Comment (ibuclaw) [inline src/dmd/dcast.d]: Why is this in dcast.d?

File: src/dmd/dsymbol.d
Comment (ibuclaw) [inline src/dmd/dsymbol.d]: Here, extra logic needlessly moved to caller.

File: src/dmd/dtemplate.d
Comment (ibuclaw) [inline src/dmd/dtemplate.d]: And here

File: src/dmd/expression.d
Comment (ibuclaw) [inline src/dmd/expression.d]: And here

File: src/dmd/expressionsem.d
Comment (ibuclaw) [inline src/dmd/expressionsem.d]: And here

File: src/dmd/expressionsem.d
Comment (ibuclaw) [inline src/dmd/expressionsem.d]: And here

File: src/dmd/expressionsem.d
Comment (ibuclaw) [inline src/dmd/expressionsem.d]: And here

File: src/dmd/expressionsem.d
Comment (ibuclaw) [inline src/dmd/expressionsem.d]: And here

File: src/dmd/expressionsem.d
Comment (ibuclaw) [inline src/dmd/expressionsem.d]: And here

File: src/dmd/expressionsem.d
Comment (ibuclaw) [inline src/dmd/expressionsem.d]: And here

File: src/dmd/opover.d
Comment (ibuclaw) [inline src/dmd/opover.d]: And here

File: src/dmd/statementsem.d
Comment (ibuclaw) [inline src/dmd/statementsem.d]: And here

File: src/dmd/statementsem.d
Comment (ibuclaw) [inline src/dmd/statementsem.d]: And finally here (may have missed a few). All noted examples are where the caller is having to handle alias this, but it really shouldn't.

File: src/dmd/dtemplate.d
Comment (ibuclaw) [inline src/dmd/dtemplate.d]: Adding `loc` parameter to `deduceType` is at least one thing you can move to a separate PR. :-)

File: test/fail_compilation/aliasthis_scope.d
Comment (ibuclaw) [inline test/fail_compilation/aliasthis_scope.d]: 1. Why is this error reported twice?
2. Why is it reporting on the wrong line?

File: src/dmd/dtemplate.d
Comment (ibuclaw) [inline src/dmd/dtemplate.d]: This pattern of locking/unlocking that's repeated in many places should not be shouldered to the caller. Consider moving to a function.

File: src/dmd/dsymbolsem.d
Comment (ibuclaw) [inline src/dmd/dsymbolsem.d]: Why has this error been downgraded?

File: src/dmd/aliasthis.d
Comment (ibuclaw) [inline src/dmd/aliasthis.d]: Introducing a new deprecation, but no tests for it?

**Review by ibuclaw — CHANGES_REQUESTED** [review]
In its current form. Too much work seems to be moved out of aliasthis to the caller. All managing of alias this resolving should have better encapsulation.

**Comment by 12345swordy** [comment]
Status?

**Comment by sighoya** [comment]
What's the state?

**Comment by jblachly** [comment]
Status update?

Just realized we really, really need this in our library.

**Comment by jacob-carlborg** [comment]
> Just realized we really, really need this (multiple alias this) in our library.

@jblachly Depending on your needs, you might get away with a library implementation relying on `opDispatch`.

**Comment by RazvanN7** [comment]
As far as I can remember, @WalterBright was against adding multiple alias this support because it leads to the diamond problem. Also, while we were discussing, he has mentioned that it was a mistake to allow classes to have an alias this. @WalterBright what are your thoughts on this?

**Comment by dkorpel** [comment]
alias this has been deprecated for classes, going in the opposite direction of this PR, so I'll close this.

**Pull request closed without merging**
